### ISSUE TITLE
Using it with DatabaseCache

### ISSUE THREAD

**Issue by aericson** [BertrandBordage/django-cachalot#56]
Is it possible to use it with DatabaseCache?

using the following config:

```
CACHES['cachalot'] = {
    'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
    'LOCATION': 'cachalot_cache_table',
}

CACHALOT_CACHE = 'cachalot'

CACHALOT_UNCACHABLE_TABLES = frozenset(('django_migrations', 'cachalot_cache_table'))
```

but I'm getting a maximum recursion depth.

```
Traceback:
File "***/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  223.                 response = middleware_method(request, response)
File "***/lib/python2.7/site-packages/django/contrib/sessions/middleware.py" in process_response
  50.                         request.session.save()
File "***/lib/python2.7/site-packages/django/contrib/sessions/backends/db.py" in save
  56.             return self.create()
File "***/lib/python2.7/site-packages/django/contrib/sessions/backends/db.py" in create
  41.                 self.save(must_create=True)
File "***/lib/python2.7/site-packages/django/contrib/sessions/backends/db.py" in save
  65.                 obj.save(force_insert=must_create, using=using)
File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py" in inner
  136.                 self.using, exc_type is None and not needs_rollback)
File "***/lib/python2.7/site-packages/cachalot/cache.py" in exit_atomic
  51.                 atomic_cache.commit()
File "***/lib/python2.7/site-packages/cachalot/transaction.py" in commit
  27.         self.parent_cache.set_many(self, None)
File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py" in set_many
  194.             self.set(key, value, timeout=timeout, version=version)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in set
  89.         self._base_set('set', key, value, timeout)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in _base_set
  143.                                        [key, b64encoded, exp])
File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py" in inner
  136.                 self.using, exc_type is None and not needs_rollback)
File "***/lib/python2.7/site-packages/cachalot/cache.py" in exit_atomic
  51.                 atomic_cache.commit()
File "***/lib/python2.7/site-packages/cachalot/transaction.py" in commit
  27.         self.parent_cache.set_many(self, None)
File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py" in set_many
  194.             self.set(key, value, timeout=timeout, version=version)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in set
  89.         self._base_set('set', key, value, timeout)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in _base_set
  143.                                        [key, b64encoded, exp])
...
File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py" in inner
  136.                 self.using, exc_type is None and not needs_rollback)
File "***/lib/python2.7/site-packages/cachalot/cache.py" in exit_atomic
  51.                 atomic_cache.commit()
File "***/lib/python2.7/site-packages/cachalot/transaction.py" in commit
  27.         self.parent_cache.set_many(self, None)
File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py" in set_many
  194.             self.set(key, value, timeout=timeout, version=version)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in set
  89.         self._base_set('set', key, value, timeout)
File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py" in _base_set
  139.                                        [b64encoded, exp, key])
File "***/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)
File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py" in inner
  112.                     invalidate(*tables, db_alias=cursor.db.alias)
File "***/lib/python2.7/site-packages/cachalot/api.py" in invalidate
  65.             _get_tables(tables_or_models), cache_alias, db_alias):
File "***/lib/python2.7/site-packages/cachalot/api.py" in _cache_db_tables_iterator
  23.             tables = connections[db_alias].introspection.table_names()
File "***/lib/python2.7/site-packages/django/db/backends/base/introspection.py" in table_names
  56.             with self.connection.cursor() as cursor:
File "***/lib/python2.7/site-packages/django/db/backends/base/base.py" in cursor
  161.         if self.queries_logged:
File "***/lib/python2.7/site-packages/django/db/backends/base/base.py" in queries_logged
  76.         return self.force_debug_cursor or settings.DEBUG

Exception Type: RuntimeError at /
Exception Value: maximum recursion depth exceeded
```

Thanks!

**Comment by aericson** [comment]
In fact this happens if I have a default cache and another cache with the `DatabaseCache` backend.

Just had the same problem with the following:
```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.locmem.LocMemCache',
        'LOCATION': 'unique-snowflake',
    }
}

CACHES['collectfast'] = {
    'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
    'LOCATION': 'collectfast_cached_table',
    'OPTIONS': {
        'MAX_ENTRIES': 2000,
    }
}

COLLECTFAST_CACHE = 'collectfast'
CACHALOT_UNCACHABLE_TABLES = frozenset(('django_migrations', 'collectfast_cached_table'))
```

and the stacktrace while running `python manage.py migrate`:

```
Running migrations:
  No migrations to apply.
Traceback (most recent call last):
  File "manage.py", line 16, in <module>
    execute_from_command_line(sys.argv)
  File "***/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "***/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "***/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "***/lib/python2.7/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "***/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 226, in handle
    emit_post_migrate_signal(created_models, self.verbosity, self.interactive, connection.alias)
  File "***/lib/python2.7/site-packages/django/core/management/sql.py", line 280, in emit_post_migrate_signal
    using=db)
  File "***/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 145, in _invalidate_on_migration
    invalidate(*sender.get_models(), db_alias=kwargs['using'])
  File "***/lib/python2.7/site-packages/cachalot/api.py", line 67, in invalidate
    cachalot_caches.get_cache(cache_alias, db_alias), db_alias, tables)
  File "***/lib/python2.7/site-packages/cachalot/utils.py", line 169, in _invalidate_tables
    {_get_table_cache_key(db_alias, t): now for t in tables}, None)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 143, in _base_set
    [key, b64encoded, exp])
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 136, in inner
    self.using, exc_type is None and not needs_rollback)
  File "***/lib/python2.7/site-packages/cachalot/cache.py", line 51, in exit_atomic
    atomic_cache.commit()
  File "***/lib/python2.7/site-packages/cachalot/transaction.py", line 27, in commit
    self.parent_cache.set_many(self, None)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 143, in _base_set
    [key, b64encoded, exp])
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 136, in inner
    self.using, exc_type is None and not needs_rollback)
  File "***/lib/python2.7/site-packages/cachalot/cache.py", line 51, in exit_atomic
    atomic_cache.commit()
  File "***/lib/python2.7/site-packages/cachalot/transaction.py", line 27, in commit
    self.parent_cache.set_many(self, None)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 143, in _base_set
    [key, b64encoded, exp])
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 136, in inner
    self.using, exc_type is None and not needs_rollback)
  File "***/lib/python2.7/site-packages/cachalot/cache.py", line 51, in exit_atomic
    atomic_cache.commit()
  File "***/lib/python2.7/site-packages/cachalot/transaction.py", line 27, in commit
    self.parent_cache.set_many(self, None)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 143, in _base_set
    [key, b64encoded, exp])
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 136, in inner
    self.using, exc_type is None and not needs_rollback)
  File "***/lib/python2.7/site-packages/cachalot/cache.py", line 51, in exit_atomic
    atomic_cache.commit()
  File "***/lib/python2.7/site-packages/cachalot/transaction.py", line 27, in commit
    self.parent_cache.set_many(self, None)

...
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 143, in _base_set
    [key, b64encoded, exp])
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 136, in inner
    self.using, exc_type is None and not needs_rollback)
  File "***/lib/python2.7/site-packages/cachalot/cache.py", line 51, in exit_atomic
    atomic_cache.commit()
  File "***/lib/python2.7/site-packages/cachalot/transaction.py", line 27, in commit
    self.parent_cache.set_many(self, None)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/base.py", line 194, in set_many
    self.set(key, value, timeout=timeout, version=version)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 89, in set
    self._base_set('set', key, value, timeout)
  File "***/lib/python2.7/site-packages/django/core/cache/backends/db.py", line 139, in _base_set
    [b64encoded, exp, key])
  File "***/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 112, in inner
    invalidate(*tables, db_alias=cursor.db.alias)
  File "***/lib/python2.7/site-packages/cachalot/api.py", line 65, in invalidate
    _get_tables(tables_or_models), cache_alias, db_alias):
  File "***/lib/python2.7/site-packages/cachalot/api.py", line 23, in _cache_db_tables_iterator
    tables = connections[db_alias].introspection.table_names()
  File "***/lib/python2.7/site-packages/django/db/backends/base/introspection.py", line 57, in table_names
    return get_names(cursor)
  File "***/lib/python2.7/site-packages/django/db/backends/base/introspection.py", line 53, in get_names
    return sorted(ti.name for ti in self.get_table_list(cursor)
  File "***/lib/python2.7/site-packages/django/db/backends/postgresql_psycopg2/introspection.py", line 53, in get_table_list
    AND pg_catalog.pg_table_is_visible(c.oid)""")
  File "***/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "***/lib/python2.7/site-packages/cachalot/monkey_patch.py", line 106, in inner
    out = original(cursor, sql, *args, **kwargs)
  File "***/lib/python2.7/site-packages/django/db/backends/utils.py", line 60, in execute
    with self.db.wrap_database_errors:
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

Thanks!

**Comment by BertrandBordage** [comment]
No, it’s not currently possible to use django-cachalot with DatabaseCache. It could be possible, but since querying the database is almost always slower than using a real cache engine, it’s not worth it.

I will probably change my mind in the future if my previous statement is no longer true.

**Comment by aericson** [comment]
@BertrandBordage what about having a DatabaseCache used for something else.

Can't get it to work now. Check my example in https://github.com/BertrandBordage/django-cachalot/issues/56#issuecomment-240951128.

**Comment by BertrandBordage** [comment]
Ah, indeed, didn’t realize that. I’m investigating it.

**Comment by aericson** [comment]
Thanks! Just let me know if you need any info.

**Comment by BertrandBordage** [comment]
This issue should be resolved by 246873b42fd747fe1d82903d0651f5d9e04e7deb. Django-cachalot was invalidating uncachable tables. This was not really a bug since I thought it was safer to invalidate uncachable tables in case someone changed the `UNCACHABLE_TABLES` or `ONLY_CACHABLE_TABLES`. But it’s not really a big problem to ask users to invalidate django-cachalot after changing these setting. Therefore, I specified it with 82262184c8ca99ce874efc7710e02b466828cd9d for the sake of clarity :)

Note that django-cachalot still does not support `DatabaseCache` because it requires some additional setup: users have to add the name of their cache table to `CACHALOT_UNCACHABLE_TABLES`, and they need to manually create the cache table using `./manage.py createcachetable`. There is also another reason for me to avoid supporting it: it’s more complex to test. Because the SQL queries executed by `DatabaseCache` interfere with the number of SQL queries I measure in tests, the tests are broken and even adjusted, it would be tougher to be sure tests are running fine.
